pages/a/ms-azure-guest-post.html

```diff
@@ -0,0 +1,89 @@
+- **Subnet**: Further segmentation for virtual networks.
+- **Virtual Machine (VM)**: Represents a particular operating system or application stack.
+- **Network Interface (NIC)**: Represents a network interface that can be associated to one or more virtual machines (VM).
+- **Network Security Group (NSG)**: Contains a list of rules that allow or deny network traffic to your VMs in a Virtual Network. Can be associated with subnets or individual VM instances within that subnet.
+
+## Data Set
+
+70 terabytes of compressed data from the [common crawl database](http://commoncrawl.org/the-data/).
+
+## Plan A
+
+Our initial approach was to create a series of scripts to do setup the following:
+
+1. A 3 node [Consul](https://www.consul.io/) cluster to manage the cluster and handle service discovery.
+2. 3 Docker swarm master nodes to create VM instances quickly via docker machine.
+3. Gradually scale the cluster by creating docker instances running Crate and managed by the Swarm master.
+4. A 3 node monitoring and metrics cluster, we settled on [Ruxit](https://ruxit.com/).
+5. Import our data set using Crate's `Copy From` command.
+
+After a two day coding session with Microsoft staff in Vienna, they recommended we instead switched to using [Azure Templates](https://azure.microsoft.com/en-us/documentation/articles/resource-group-authoring-templates/), a json formatted file that provisions all the resources and commands to be run for your application in a single operation. These templates can be used in conjunction with the Azure command line tool to rigger a complex cluster quickly, for example:
+
+```bash
+azure group deployment create -f azuredeploy.json \
+  -e azuredeploy.parameters.json \
+  -g <resourcegroup>
+```
+
+For the convenience of our team and to allow for reproduction of individual clusters, we broke these templates into cluster functions. I.e. one template for creating the network infrastructure, one for installing Docker components etc.
+
+### Problems
+
+Plan A started well, but after a few hundred nodes we started to run into problems from different sources. First the network overhead of running traffic through the [Docker container bridge](https://docs.docker.com/engine/userguide/networking/dockernetworks/#a-bridge-network) proved to much and the Microsoft engineers recommended we switched to running Crate from Linux packages. Next there were problems with the Consul cluster as we had not configured it optimally to run on a multi-core machine and so it was struggling to meet the demand put upon it. After many hours debugging leading to a small config change we were able to get Consul running using all cores and reacting far more responsively. Finally despite being a great service, the Ruxit service couldn't cope with the traffic we were generating and the 29 metrics we wanted to collect. At around 500 nodes we were hitting their API so hard it shut down, so this wasn't a viable option for us, or for Ruxit.
```

Comment:
we should also mention, why Ruxit cannot keep up this large scale:
the reason is that they monitor each process and the connections between processes, which leads in our case to a massive amount of data. @joemoe knows more about it